File: stack_cc/practice.cc
#include "code.h"

// This is a practice file.  Delete methods from this file, then rewrite them
// and run the tests by hitting <c-j>.  That way you don't have to mess with
// the reference implementation.

Stack::Stack() {
    head = NULL;
    return;
}

Stack::~Stack() {
    Node *next;
    while (head) {
        next = head->nxt;
        delete head;
        head = next;
    }
    return;
}

void Stack::push(int value) {
    Node *node = new Node;
    node->value = value;
    node->nxt = head;
    head = node;
    return;
}

int Stack::pop() {
    Node *poppednode;
    if (head == NULL) {
        throw "you can't do that";
    }
    poppednode = head;
    int value = poppednode->value;
    head = poppednode->nxt;
    delete poppednode;
    return value;
}

File: data_structures/stack/stack_rb/practice.rb
# This is a practice file.  Delete methods from this file, then rewrite them
# and run the tests by hitting <c-j>.  That way you don't have to mess with
# the reference implementation.

class Node
  attr_accessor :nxt, :value

  def initialize(value)
    @value = value
    @nxt = nil
  end
end


class Stack

  attr_reader :head

  def initialize()
    @head = nil 
  end

  def push(value)
    node = Node.new(value)
    if @head != nil
      node.nxt = @head
    end
    @head = node
  end

  def pop
    if @head == nil
      return nil
    else
      old = @head
      @head = @head.nxt
      value = old.value
      # [tag:todo:gem] Need to delete "old"
      return value
    end
  end

end

# vim:set ft=ruby sw=2 sts=2 et:
